### ISSUE TITLE
Error with destructuring defaults in async function default arguments

### ISSUE THREAD

**Issue by jimmydief** [rpetrich/babel-plugin-transform-async-to-promises#29]
While trying out the default arguments changes in 0.8.10, I found one more issue that presented itself when running with `babel-plugin-istanbul`. 

```js
let x = 0, z = 0;

async function f({y = (z++, {})} = (x++, {})) {}
```

I get this error:
```js
~/node_modules/@babel/types/lib/definitions/utils.js:128
      throw new TypeError(`Property ${key} of ${node.type} expected node to be of a type ${JSON.stringify(types)} ` + `but instead got ${JSON.stringify(val && val.type)}`);

TypeError: Property left of BinaryExpression expected node to be of a type ["Expression"] but instead got "ObjectPattern"
    at Object.validate (~/node_modules/@babel/types/lib/definitions/utils.js:128:13)
    at validate (~/node_modules/@babel/types/lib/validators/validate.js:17:9)
    at builder (~/node_modules/@babel/types/lib/builders/builder.js:46:27)
    at Object.BinaryExpression (~/node_modules/@babel/types/lib/builders/generated/index.js:247:31)
    at rewriteDefaultArguments (~/node_modules/babel-plugin-transform-async-to-promises/async-to-promises.js:3129:43)
    at PluginPass.ClassMethod (~/node_modules/babel-plugin-transform-async-to-promises/async-to-promises.js:3248:25)
    at newFn (~/node_modules/@babel/traverse/lib/visitors.js:193:21)
    at NodePath._call (~/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (~/node_modules/@babel/traverse/lib/path/context.js:40:17)
    at NodePath.visit (~/node_modules/@babel/traverse/lib/path/context.js:88:12)
```

Seems like default argument rewriting needs to consider destructuring defaults when the right-hand side of the default assignment is a complex expression.

**Comment by orlov-vo** [comment]
I have another error, which connected with default value in argument:
```
TypeError: Property init of VariableDeclarator expected node to be of a type ["Expression"] but instead got "ObjectPattern"
```

with this code:
```js
let result;
export default function foo( objectOrFunction = {} ) {
  return async () => {
    result = typeof objectOrFunction === "function" ? await objectOrFunction() : objectOrFunction;
  };
}
```

**Comment by rpetrich** [comment]
Thanks for reporting this bug. I've imported it as a test case and released a fix in 0.8.11
